lib/public_key/src/pubkey_cert.erl

```diff
@@ -144,15 +144,29 @@ validate_time(OtpCert, UserState, VerifyFun) ->
     {'Validity', NotBeforeStr, NotAfterStr} 
 	= TBSCert#'OTPTBSCertificate'.validity,
     Now = calendar:datetime_to_gregorian_seconds(calendar:universal_time()),
-    NotBefore = time_str_2_gregorian_sec(notBefore, NotBeforeStr),
-    NotAfter = time_str_2_gregorian_sec(notAfter, NotAfterStr),
 
-    case ((NotBefore =< Now) and (Now =< NotAfter)) of
-	true ->
-	    UserState;
-	false ->
-	    verify_fun(OtpCert, {bad_cert, cert_expired}, UserState, VerifyFun)
+    % Catch invalid date format errors in the `validity` field of the certificate
+    NotExpired = try
+        NotBefore = time_str_2_gregorian_sec(notBefore, NotBeforeStr),
+        NotAfter = time_str_2_gregorian_sec(notAfter, NotAfterStr),
+        ((NotBefore =< Now) and (Now =< NotAfter))
+        % "error:function_clause" is thrown by time_str_2_gregorian_sec if the date format is not valid
+        % verify_fun only throws exceptions
+    catch error:function_clause ->
+        verify_fun(OtpCert, {bad_cert, invalid_validity_dates}, UserState, VerifyFun)
+    end,
+
+    case NotExpired of
+        true ->
+            UserState;
+        false ->
+            verify_fun(OtpCert, {bad_cert, cert_expired}, UserState, VerifyFun);
+            % This will be called in case an exception is caught when processing dates
+            % and verify_fun alters the state, so we keep the modified state along the way.
+        NewState ->
+            NewState
```

Comment:
I have doubts about this catch everything clause.
Previous code didn't work that way.

Is there a reason for it?

Reply:
No reason, preventing crash in case of unexpected value. I can remove it if necessary.

Reply:
I actually think that the last case is the case that the catch verify_fun returns a state.  I think the following code could benefit readablility:

``validate_time(OtpCert, UserState, VerifyFun) ->
    TBSCert = OtpCert#'OTPCertificate'.tbsCertificate,
    {'Validity', NotBeforeStr, NotAfterStr} 
	= TBSCert#'OTPTBSCertificate'.validity,
   Now = calendar:datetime_to_gregorian_seconds(calendar:universal_time()),

  case valid_dates(NotBeforeStr, NotAfterStr}) of
   {true, {NotBefore, NotAfter}} ->
         case ((NotBefore =< Now) and (Now =< NotAfter)) of
	    true ->
	      UserState;
	   false ->
	      verify_fun(OtpCert, {bad_cert, cert_expired}, UserState, VerifyFun)
           end;
   {false, Error} ->
        verify_fun(OtpCert, {bad_cert, Error}, UserState, VerifyFun)
  end.

valid_dates(NotBeforeStr, NotAfterStr) ->
  try
    NotBefore = time_str_2_gregorian_sec(notBefore, NotBeforeStr),
    NotAfter = time_str_2_gregorian_sec(notAfter, NotAfterStr),
    {true, {NotBefore, NotAfter}}
catch
    error:function_clause ->
    {false, invalid_validity_dates}
end.